language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddConditionalDefaultValueCodeAction.java

```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Override
+    public List<CodeAction> getCodeActions(Diagnostic diagnostic,
+                                           DiagBasedPositionDetails positionDetails,
+                                           CodeActionContext context) {
+        Optional<TypeSymbol> actualTypeSymbol = positionDetails.diagnosticProperty(
+                DiagBasedPositionDetails.DIAG_PROP_INCOMPATIBLE_TYPES_FOUND_SYMBOL_INDEX);
+        Optional<TypeSymbol> expectedTypeSymbol = positionDetails.diagnosticProperty(
+                DiagBasedPositionDetails.DIAG_PROP_INCOMPATIBLE_TYPES_EXPECTED_SYMBOL_INDEX);
+
+        if (expectedTypeSymbol.isEmpty() || actualTypeSymbol.isEmpty()
+                || !isOptionalType(actualTypeSymbol.get(), context)) {
+            return Collections.emptyList();
+        }
+        Optional<TypeSymbol> typeWithoutOptional = getTypeWithoutOptional(actualTypeSymbol.get(), context);
+        if (typeWithoutOptional.isEmpty() || !typeWithoutOptional.get().subtypeOf(expectedTypeSymbol.get())) {
+            return Collections.emptyList();
+        }
+        Optional<String> defaultValue = DefaultValueGenerationUtil.getDefaultValueForType(typeWithoutOptional.get());
+        if (defaultValue.isEmpty()) {
+            return Collections.emptyList();
+        }
+        Node matchedNode = positionDetails.matchedNode();
+        Position endPosition = PositionUtil.toPosition(matchedNode.lineRange().endLine());
+        String editText = " ?: " + defaultValue.get();
+        TextEdit textEdit = new TextEdit(new Range(endPosition, endPosition), editText);
+        return Collections.singletonList(CodeActionUtil.createCodeAction(CommandConstants.ADD_CONDITIONAL_DEFAULT,
+                List.of(textEdit), context.fileUri(), CodeActionKind.QuickFix));
+    }
+
+    private Optional<TypeSymbol> getTypeWithoutOptional(TypeSymbol actualType, CodeActionContext context) {
+        Optional<SemanticModel> semanticModel = context.currentSemanticModel();
+        if (actualType.typeKind() != TypeDescKind.UNION || semanticModel.isEmpty()) {
+            return Optional.empty();
+        }
+        List<TypeSymbol> memberTypes = ((UnionTypeSymbol) actualType).memberTypeDescriptors().stream()
+                .filter(typeSymbol -> typeSymbol.typeKind() != TypeDescKind.NIL).collect(Collectors.toList());
+        UnionTypeSymbol unionType = semanticModel.get().types().builder().UNION_TYPE
+                .withMemberTypes(memberTypes.toArray(TypeSymbol[]::new)).build();
+        return Optional.of(unionType);
+    }
+
+    private boolean isOptionalType(TypeSymbol typeSymbol, CodeActionContext context) {
+        Optional<SemanticModel> semanticModel = context.currentSemanticModel();
+        if (semanticModel.isEmpty()) {
+            return false;
+        }
+        TypeSymbol nilType = semanticModel.get().types().NIL;
+        return nilType.subtypeOf(typeSymbol);
+    }
```

Comment:
We can move the `sementicModel.isPresent()` check to the validate() and make this check, a one liner in line 72 `context.currentSemanticModel().get().types().NIL.subtypeOf(actualTypeSymbol.get()`. But only if the code readability is not affected.